lib/datasource/deb/index.spec.ts

```diff
@@ -0,0 +1,189 @@
+import { copyFile, mkdirp, remove, stat } from 'fs-extra';
+import { getPkgReleases } from '..';
+import type { GetPkgReleasesConfig } from '..';
+import * as httpMock from '../../../test/http-mock';
+import { GlobalConfig } from '../../config/global';
+import { DebDatasource } from '.';
+
+describe('datasource/deb/index', () => {
+  describe('getReleases', () => {
+    const testPackagesFile = __dirname + '/test-data/Packages.gz';
+    const extractedTestFile = __dirname + '/test-data/Packages';
+    const cacheDir = '/tmp/renovate-cache/';
+    // const downloadFolder =
+    //   cacheDir + 'others/' + DebDatasource.downloadDirectory + '/';
+    const extractionFolder =
+      cacheDir + 'others/' + DebDatasource.cacheSubDir + '/';
+    // const compressedPackageFile =
+    //   downloadFolder +
+    //   '0b01d9df270158d22c09c85f21b0f403d31b0da3cae4930fdb305df8f7749c27.gz';
+    const extractedPackageFile =
+      extractionFolder +
+      '0b01d9df270158d22c09c85f21b0f403d31b0da3cae4930fdb305df8f7749c27.txt';
+
+    GlobalConfig.set({ cacheDir: cacheDir });
+
+    let cfg: GetPkgReleasesConfig; // this can be modified within the test cases
+
+    beforeEach(async () => {
+      jest.resetAllMocks();
+      await remove(cacheDir);
+      cfg = {
+        datasource: 'deb',
+        depName: 'steam-devices',
+        registryUrls: [
+          'http://ftp.debian.org/debian?suite=stable&components=non-free&binaryArch=amd64',
+        ],
+      };
+    });
+
+    it('returns a valid version for the package `steam-devices` and does not require redownload', async () => {
+      // copy the Packages file to the appropriate location
+      await mkdirp(extractionFolder);
+      await copyFile(extractedTestFile, extractedPackageFile);
+      const stats = await stat(extractedPackageFile);
+      const ts = stats.ctime;
+
+      httpMock
+        .scope('http://ftp.debian.org')
+        .head('/debian/dists/stable/non-free/binary-amd64/Packages.gz')
+        .reply(304);
+
+      const res = await getPkgReleases(cfg);
+      expect(res).toBeObject();
+      expect(res.releases).toHaveLength(1);
+
+      // validate that the server was called correctly
+      expect(httpMock.getTrace()).toHaveLength(1);
```

Comment:
Unfulfilled mocks are only evaluated after the test block. In the next line the array access on `getTrace()` would therefore be executed which would lead to an runtime exception in a failing test case scenario.

Reply:
Yes and no, if you look at the http mock code, you can see a `afterEach` which will validate this, so we don't have to repeat it here. 😉